subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/recomp/CompilationSourceDirs.java

```diff
@@ -38,32 +42,35 @@
 @NonNullApi
 public class CompilationSourceDirs {
     private static final Logger LOG = LoggerFactory.getLogger(CompilationSourceDirs.class);
+    private final List<File> sourceRoots;
 
-    private final FileTreeInternal sources;
-    private SourceRoots sourceRoots;
-
-    public CompilationSourceDirs(FileTreeInternal sources) {
-        this.sources = sources;
-    }
-
-    public List<File> getSourceRoots() {
-        return resolveRoots().getSourceRoots();
+    public CompilationSourceDirs(List<File> sourceRoots) {
+        this.sourceRoots = sourceRoots;
     }
 
-    public boolean canInferSourceRoots() {
-        return resolveRoots().isCanInferSourceRoots();
+    public static List<File> inferSourceRoots(FileTreeInternal sources) {
+        SourceRoots visitor = new SourceRoots();
+        sources.visitLeafCollections(visitor);
+        return visitor.canInferSourceRoots ? visitor.sourceRoots : Collections.emptyList();
```

Comment:
Use the getters for `canInferSourceRoots` and `sourceRoots` here, so they are used.

Reply:
Removed public getters because nobody is using them outside the class.